File: sdks/go/manifest.go
package aegis

import (
	"fmt"
	"regexp"
	"strings"
)

const ProtocolVersion = "1.0.0"

type Checkpoint struct { Cursor string; Watermark string }
type RetryDirective struct { RetryAfterMs int; Reason string }

func RetryForStatus(status int, retryAfterMs int) *RetryDirective {
	if status != 429 && status != 502 && status != 503 && status != 504 { return nil }
	if retryAfterMs < 250 { retryAfterMs = 250 }
	reason := "transient_failure"; if status == 429 { reason = "rate_limit" }
	return &RetryDirective{RetryAfterMs: retryAfterMs, Reason: reason}
}

type Manifest struct {
	ProtocolVersion string
	ID              string
	Capabilities    []string
	MinimumScopes   []string
}

var connectorID = regexp.MustCompile(`^[a-z][a-z0-9-]{2,62}$`)

func ValidateReadOnlyManifest(manifest Manifest) error {
	if manifest.ProtocolVersion != ProtocolVersion {
		return fmt.Errorf("unsupported connector protocol version")
	}
	if !connectorID.MatchString(manifest.ID) {
		return fmt.Errorf("invalid connector ID")
	}
	if len(manifest.Capabilities) == 0 || len(manifest.MinimumScopes) == 0 {
		return fmt.Errorf("read-only connector manifests require capabilities and scopes")
	}
	for _, capability := range manifest.Capabilities {
		if capability != "IDENTITY_READ" && capability != "ACCESS_GRAPH_READ" && capability != "USAGE_READ" {
			return fmt.Errorf("capability %s is not read-only", capability)
		}
	}
	return nil
}

func RedactHeaders(headers map[string]string) map[string]string {
	redacted := make(map[string]string, len(headers))
	for key, value := range headers {
		if strings.Contains(strings.ToLower(key), "authorization") || strings.Contains(strings.ToLower(key), "token") {
			redacted[key] = "REDACTED"
		} else {
			redacted[key] = value
		}
	}
	return redacted
}

File: sdks/python/aegis_connector_sdk/manifest.py
import re

PROTOCOL_VERSION = "1.0.0"
_READ_ONLY = {"IDENTITY_READ", "ACCESS_GRAPH_READ", "USAGE_READ"}


def checkpoint(cursor=None, watermark=None):
    return {key: value for key, value in {"cursor": cursor, "watermark": watermark}.items() if value}


def retry_for_status(status, retry_after_seconds=None):
    if status not in {429, 502, 503, 504}:
        return None
    return {"reason": "rate_limit" if status == 429 else "transient_failure", "retryAfterMs": int(max(0.25, retry_after_seconds or 0.25) * 1000)}


def validate_read_only_manifest(manifest: dict) -> None:
    if manifest.get("protocolVersion") != PROTOCOL_VERSION:
        raise ValueError("unsupported connector protocol version")
    if not re.fullmatch(r"[a-z][a-z0-9-]{2,62}", manifest.get("id", "")):
        raise ValueError("invalid connector ID")
    capabilities = set(manifest.get("capabilities", []))
    if not capabilities or not capabilities <= _READ_ONLY or not manifest.get("minimumScopes"):
        raise ValueError("read-only connector manifests require read capabilities and scopes")


def redact_fixture(value):
    if isinstance(value, list):
        return [redact_fixture(item) for item in value]
    if isinstance(value, dict):
        return {
            key: "REDACTED" if re.search(r"token|secret|authorization|password", key, re.I) else redact_fixture(item)
            for key, item in value.items()
        }
    if isinstance(value, str):
        return re.sub(r"bearer\s+\S+", "REDACTED", value, flags=re.I)
    return value
